### PR TITLE
Add assumption-based unsat cores.

### PR DESCRIPTION
**Pull request by mpreiner**
This PR adds an assumption-based unsat cores option. If enabled it will
disable proof logging in the SAT solver and adds input assertions as
assumptions to the SAT solver. When an unsat core is requested we
extract the unsat core in terms of the unsat assumption in the SAT
solver. Assumption-based unsat cores use the proof infrastructure to map
the input assumptions back to the original assertions.

### CODE REVIEW

File: src/theory/booleans/proof_checker.cpp [hunk 1]
```diff
@@ -74,6 +74,7 @@ void BoolProofRuleChecker::registerTo(ProofChecker* pc)
   pc->registerChecker(PfRule::CNF_ITE_NEG1, this);
   pc->registerChecker(PfRule::CNF_ITE_NEG2, this);
   pc->registerChecker(PfRule::CNF_ITE_NEG3, this);
+  pc->registerChecker(PfRule::SAT_REFUTATION, this);
```
Comment (ajreynol) [inline src/theory/booleans/proof_checker.cpp:77]:
```suggestion
  pc->registerTrustedChecker(PfRule::SAT_REFUTATION, this, 1);
```

File: src/smt/smt_engine.cpp [hunk 2]
```diff
@@ -1419,9 +1419,20 @@ UnsatCore SmtEngine::getUnsatCoreInternal()
   // generate with new proofs
   PropEngine* pe = getPropEngine();
   Assert(pe != nullptr);
-  Assert(pe->getProof() != nullptr);
-  std::shared_ptr<ProofNode> pfn = d_pfManager->getFinalProof(
-      pe->getProof(), *d_asserts, *d_definedFunctions);
+
+  std::shared_ptr<ProofNode> pepf;
+  if (options::unsatCores()
+      && options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS)
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp:1425]:
```suggestion
  if (options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS)
```

File: src/prop/prop_engine.cpp [hunk 3]
```diff
@@ -260,7 +261,19 @@ void PropEngine::assertInternal(
   // Assert as (possibly) removable
   if (isProofEnabled())
   {
-    d_pfCnfStream->convertAndAssert(node, negated, removable, pg);
+    if (options::unsatCores()
+        && options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS
```
Comment (ajreynol) [inline src/prop/prop_engine.cpp:265]:
```suggestion
    if (options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS
```

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few minor changes implied by recent refactoring to options dependencies.

File: src/smt/smt_engine.cpp (hunk 2 above)
Comment (HanielB) [reply src/smt/smt_engine.cpp:1425]: After this change I'd actually inline the ite, but as you wish.

File: src/prop/prop_engine.cpp [hunk 4]
```diff
@@ -626,5 +653,29 @@ std::shared_ptr<ProofNode> PropEngine::getProof()
...
+  for (const SatLiteral& lit : unsat_assumptions)
+  {
+    core.push_back(d_cnfStream->getNode(lit));
+  }
+}
+
+std::shared_ptr<ProofNode> PropEngine::getRefutation()
+{
+  Assert(options::unsatCores());
+  Assert(options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS);
```
Comment (HanielB) [inline src/prop/prop_engine.cpp:671]:
```suggestion
  Assert(options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS);
```

File: src/prop/prop_engine.cpp [hunk 5]
```diff
@@ -626,5 +653,29 @@ std::shared_ptr<ProofNode> PropEngine::getProof()
 
 bool PropEngine::isProofEnabled() const { return d_pfCnfStream != nullptr; }
 
+void PropEngine::getUnsatCore(std::vector<Node>& core)
+{
+  Assert(options::unsatCores());
+  Assert(options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS);
```
Comment (HanielB) [inline src/prop/prop_engine.cpp:659]:
```suggestion
  Assert(options::unsatCoresMode() == options::UnsatCoresMode::ASSUMPTIONS);
```

File: src/prop/minisat/minisat.cpp [hunk 6]
```diff
@@ -193,6 +193,40 @@ SatValue MinisatSatSolver::solve() {
...
+
+  SatValue result = toSatLiteralValue(d_minisat->solve(assumps));
+  d_minisat->clearInterrupt();
+  return result;
+}
+
+void MinisatSatSolver::getUnsatAssumptions(
+    std::vector<SatLiteral>& unsat_assumptions)
+{
+  for (size_t i = 0; i < (size_t)d_minisat->d_conflict.size(); ++i)
```
Comment (HanielB) [inline src/prop/minisat/minisat.cpp:219]: does the size need to be updated at every iteration? Otherwise cache the size.

**Review by HanielB — DISMISSED** [review]
A few minor points

File: src/smt/smt_engine.cpp (hunk 2 above)
Comment (mpreiner) [reply src/smt/smt_engine.cpp]: Doesn't really make it more readable, I'll keep the if/else.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**